File: ch6/akkademy-db-java/src/main/java/com/akkademy/ClusterController.java
package com.akkademy;

import akka.actor.AbstractActor;
import akka.cluster.Cluster;
import akka.event.Logging;
import akka.event.LoggingAdapter;
import akka.japi.pf.ReceiveBuilder;

import static akka.cluster.ClusterEvent.*;

public class ClusterController extends AbstractActor {
    protected final LoggingAdapter log = Logging.getLogger(context().system(), this);
    Cluster cluster = Cluster.get(getContext().system());

    @Override
    public void preStart() {
        cluster.subscribe(self(), initialStateAsEvents(),
                MemberEvent.class, UnreachableMember.class);
    }

    @Override
    public void postStop() {
        cluster.unsubscribe(self());
    }

    private ClusterController(){
        receive(ReceiveBuilder.
                        match(MemberEvent.class, message -> {
                            log.info("MemberEvent: {}", message);
                        }).
                        match(UnreachableMember.class, message -> {
                            log.info("UnreachableMember: {}", message);
                        }).build()
        );
    }
}

File: ch6/akkademaid-scala/src/main/scala/com/akkademy/ClusterController.scala
package com.akkademy

import akka.actor.Actor
import akka.cluster.Cluster
import akka.cluster.ClusterEvent.{UnreachableMember, MemberEvent}
import akka.event.Logging

class ClusterController extends Actor {
  val log = Logging(context.system, this)
  val cluster = Cluster(context.system)

  override def preStart() {
    cluster.subscribe(self, classOf[MemberEvent], classOf[UnreachableMember])
  }

  override def postStop() {
    cluster.unsubscribe(self)
  }

  override def receive = {
    case x: MemberEvent => log.info("MemberEvent: {}", x)
    case x: UnreachableMember => log.info("UnreachableMember {}: ", x)
  }
}

